### PR TITLE
Improve ControlFlow analysis

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR makes improvements to the controlflow analysis, and particularly to the Region analysis:

- Fix a couple of bugs
- Add documentation
- Add tests
- make region building more generic for testing
- add invariant assertions
- disable loop support (it was not supported anyways, and need additional work to work properly)

### CODE REVIEW

File: chiselc/src/tools/functions.rs [hunk 1]
```diff
@@ -19,7 +24,16 @@ impl<'a> ArrowFunction<'a> {
         match &arrow.body {
             BlockStmtOrExpr::BlockStmt(block) => {
                 let (cfg, stmt_map) = ControlFlow::build(&block.stmts)?;
-                let regions = Region::from_cfg(&cfg, &stmt_map);
+                println!("{}", Dot::new(&cfg.inner));
```
Comment (glommer) [inline chiselc/src/tools/functions.rs:27]: extraneous print ?

Comment (MarinPostma) [reply chiselc/src/tools/functions.rs:27]: Ah yes! good catch!

File: chiselc/src/tools/analysis/region.rs [hunk 2]
```diff
@@ -2,39 +2,52 @@
...
+    Loop,
+    /// Basic block component, like an expression statement, a return statement, a variable
+    /// declaration...
+    BBComponent,
+    /// A ghost node: does not give any information about controlflow
+    Ignore,
+}
+
+/// A conditional region is comprised of three regions: the test region, contains the test
+/// expression, the cons region contains the regions evaluated on true, and the alt region, the
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply chiselc/src/tools/analysis/region.rs:26]: I take this from literature. https://en.wikipedia.org/wiki/Consequent#:~:text=A%20consequent%20is%20the%20second,consequent%20is%20called%20the%20apodosis.

File: chiselc/src/tools/functions.rs [hunk 3]
```diff
@@ -19,7 +23,15 @@ impl<'a> ArrowFunction<'a> {
         match &arrow.body {
             BlockStmtOrExpr::BlockStmt(block) => {
                 let (cfg, stmt_map) = ControlFlow::build(&block.stmts)?;
-                let regions = Region::from_cfg(&cfg, &stmt_map);
+                let regions = Region::from_cfg(&cfg, &|idx| match stmt_map[idx].stmt {
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply chiselc/src/tools/functions.rs:26]: Tests mostly: I want to generate arbitrary control flows graphs, and build regions on them. This is a lot easier than writing corresponding js code

File: chiselc/src/tools/analysis/region.rs [hunk 4]
```diff
@@ -118,140 +131,620 @@ impl From<BasicBlock> for Region {
...
+            alt_region: alt_region.unwrap_or_else(|| BasicBlock::default().into()),
+        };
+
+        (Some(region.into()), cons_end)
+    }
 
-fn is_leader(idx: Idx, graph: &Graph<Node, Edge>) -> bool {
-    // is this the first statement?
-    matches!(
+    fn is_leader(&self, idx: Idx) -> bool {
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply chiselc/src/tools/analysis/region.rs:238]: will do

File: chiselc/src/tools/analysis/control_flow.rs [hunk 5]
```diff
@@ -103,16 +104,16 @@ impl<'a> CFGBuilder<'a, CfgGraph> {
...
+            // Stmt::While(while_stmt) => {
+            //     let root = self.add_stmt_node(stmt);
+            //     let (body_root, body_outs) = self.stmt(&while_stmt.body)?;
+            //     self.inner.add_edge(root, body_root, Edge::False);
+            //
+            //     self.merge_out(&body_outs, root);
+            //
+            //     Ok((root, vec![(root, Edge::True)]))
+            // }
+            Stmt::Expr(_) | Stmt::Decl(_) | Stmt::Empty(_) => {
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply chiselc/src/tools/analysis/control_flow.rs:116]: this is coming from SWC, so I simply propagate.

But in CFG it also makes sense to sometimes add empty nodes on [critical edges](https://en.wikipedia.org/wiki/Control-flow_graph#Special_edges). I think I handle that somewhere 🤔

**Pull request merged**